### PR TITLE
Feature multi threaded

### LINKED ISSUE #11: Has this been tested using goroutines?
Has this been tested with goroutines to improve speed on larger swarms?

I have used Shepherd but it turned out to be too slow, going one-by-one, which looks like this does the same. However implementing go routines with a wait group would be a good way to update larger swarms quicker.

My example is that I have a swarm where there is 22 services, each service is the same image with a slightly different configuration and traefik labels. Shepherd takes about 15 minutes to update all those services. Thinking this could handle it much quicker with the use of goroutines?

In my testing it is possible to run updates concurrently, for different services.

### PR DESCRIPTION
**Pull request by surdaft**
- Implement multi-threaded updates for faster updates in larger stacks
  - max-threads cli arg and env var

Note: I am unsure the purpose of the serviceID, if that is required then please let me know

closes #11

### CODE REVIEW

**Comment by codestation** [comment]
> Note: I am unsure the purpose of the serviceID, if that is required then please let me know

The serviceID is used to identify itself so is the last service to be updated (else the updater would be interrupted before it finishes).

I'll probably have time to check this PR on the weekend to merge it.

**Comment by surdaft** [comment]
> The serviceID is used to identify itself so is the last service to be updated (else the updater would be interrupted before it finishes).

Oh that makes absolute sense! Sausage me for not noticing. I will implement that in a moment and make sure that it doesn't update itself until the end.

**Comment by surdaft** [comment]
Just a note, I have another branch which I will later raise a PR for which adds `shoutrrr` notifications to the project. Is this something you would be ok with? Or rather I leave in my own build for now.

I am using the shoutrrr notifications version in my production environment at the moment. :)

**Comment by Rush** [comment]
I tested your branch with 10 threads on my dev machine and it seems to be working as expected. Will give it a roll in a bigger cluster.

File: swarm.go [hunk 1]
```diff
@@ -160,59 +164,94 @@ func (c *Swarm) UpdateServices(ctx context.Context, imageName ...string) error {
...
+		wg.Add(1)
+		go func(key int) {
+			defer wg.Done()
+			log.Printf("starting thread %d", key)
 
-						break
-					}
+			for {
+				if len(serviceQueue) == 0 {
+					log.Printf("exiting thread %d", key)
```
Comment (Rush) [inline swarm.go:201]: IMHO This should only be printed in debug mode.

File: swarm.go [hunk 2]
```diff
@@ -160,59 +164,94 @@ func (c *Swarm) UpdateServices(ctx context.Context, imageName ...string) error {
...
-				serviceID = service.ID
+	// only create as many as we need
+	threads := c.MaxThreads
+	if len(serviceQueue) < threads {
+		threads = len(serviceQueue)
+	}
 
-				continue
-			}
+	log.Printf("starting %d threads", threads)
```
Comment (Rush) [inline swarm.go:191]: This should only be printed in debug mode

File: swarm.go [hunk 3]
```diff
@@ -160,59 +164,94 @@ func (c *Swarm) UpdateServices(ctx context.Context, imageName ...string) error {
...
-				if ctx.Err() == context.Canceled {
-					log.Printf("Service update canceled")
+				if err = c.updateService(ctx, service); err != nil {
+					if ctx.Err() == context.Canceled {
+						log.Printf("Service update canceled")
+						return
+					}
 
-					break
+					log.Printf("Cannot update service %s: %s", service.Spec.Name, err.Error())
```
Comment (Rush) [inline swarm.go:235]: This should only be printed in debug mode

File: swarm.go [hunk 4]
```diff
@@ -90,6 +92,8 @@ func (c *Swarm) serviceList(ctx context.Context) ([]swarm.Service, error) {
 }
 
 func (c *Swarm) updateService(ctx context.Context, service swarm.Service) error {
+	log.Printf("updating service %s", service.Spec.Name)
```
Comment (Rush) [inline swarm.go:95]: Should only be printed in Debug mode

**Review by Rush — CHANGES_REQUESTED** [review]

File: swarm.go [hunk 5]
```diff
@@ -160,59 +164,94 @@ func (c *Swarm) UpdateServices(ctx context.Context, imageName ...string) error {
 		return fmt.Errorf("failed to get service list: %w", err)
 	}
 
-	var serviceID string
+	log.Printf("found %d services", len(services))
```
Comment (Rush) [inline swarm.go:167]: Same for this

**Comment by Rush** [comment]
Ok, I tested it in a larger cluster and it is broken. If you have more services for update than the threads, it will regularly skip services.

The workaround is to set threads to some insane number like 50 :-P Or maybe 1000

File: swarm.go (hunk 1 above)
Comment (surdaft) [reply swarm.go:201]: That is a very fair point, I will make those changes

**Pull request closed without merging**